File: matlab/drivers/wind_adj.m
clear all
clc

%% ACES Update to MATLAB
%-------------------------------------------------------------
% Driver for Windspeed Adjustment and Wave Growth (page 1-1 of ACES User's
% Guide). Provide estimates for wave growth over open-water and restricted
% fetches in deep and shallow water.

% Updated by: Mary Anderson, USACE-CHL-Coastal Processes Branch
% Date Created: April 28, 2011
% Date Modified:

% Requires the following functions:
% ERRWAVBRK1
% WADJ
% WAGEOS
% WAPBL
% WAPSI
% WASBL
% WASHR
% WGDL
% WGFD
% WGFET
% WGFL
% WGRO

% MAIN VARIABLE LIST:
%   INPUT
%   zobs: elevation of observed winds [m]
%   uobs: observed wind speed [m/s]
%   dtemp: air-sea temperature difference [deg C]
%   duro: duration of observed wind [hr]
%   durf: duration of final wind [hr]
%   lat: latitude of wind observation [deg]
%   windobs: wind observation type
%   fetchopt: wind fetch options
%   wgtyp: open water wave growth equation options

%   OPEN-WATER VARIABLES
%   F: length of wind fetch [m]
%   d: average depth of fetch (only for shallow water equations) [m]

%   RESTRICTED VARIABLES
%   wdir: wind direction [deg]
%   dang: radial angle increment [deg]
%   ang1: direction of first radial fetch [deg]
%   angs: fetch length [m]

%   OUTPUT
%   ue: equivalent neutral wind speed [m/s]
%   ua: adjusted wind speed [m/s]
%   Hmo: wave height [m]
%   Tp: peak wave period [s]
%   wg: type of wave-growth
%   theta: wave direction with respect to N [deg]

%   OTHERS
%-------------------------------------------------------------

SET_PATHS();

[single_case] = USER_INPUT_SINGLE_MULTI_CASE();

[metric, g, labelUnitDist, labelUnitWt] = USER_INPUT_METRIC_IMPERIAL();

if metric
    labelSpeed = 'm/s';
    labelUnitDistLrg = 'km';
else
    labelSpeed = 'mph';
    labelUnitDistLrg = 'mi';
end

windObsList = {'Overwater (shipboard)';...
    'Overwater (not shipboard)';...
    'Shore (windward - offshore to onshore)';...
    'Shore (leeward - onshore to offshore)';...
    'Over land';...
    'Geostrophic wind'};
[windobs] = USER_INPUT_FINITE_CHOICE(...
    ['Wind observation types: \n'...
        '[1] ' windObsList{1} '\n'...
        '[2] ' windObsList{2} '\n'...
        '[3] ' windObsList{3} '\n'...
        '[4] ' windObsList{4} '\n'...
        '[5] ' windObsList{5} '\n'...
        '[6] ' windObsList{6} '\n'...
        'Select option: '],...
    {'1', '2', '3', '4', '5', '6'});
windobs = str2num(windobs);

wgTypeList = {'Open Water - Deep';...
    'Open Water - Shallow';...
    'Restricted - Deep';...
    'Restricted - Shallow'};
[wgtyp] = USER_INPUT_FINITE_CHOICE(...
    ['Wind fetch and wave growth options: \n'...
        '[1] ' wgTypeList{1} '\n'...
        '[2] ' wgTypeList{2} '\n'...
        '[3] ' wgTypeList{3} '\n'...
        '[4] ' wgTypeList{4} '\n'...
        'Select option: '],...
    {'1', '2', '3', '4'});
wgtyp = str2num(wgtyp);
is_water_open = wgtyp == 1 || wgtyp == 2;
is_water_shallow = wgtyp == 2 || wgtyp == 4;
%is_water_restricted = wgtyp == 3 || wgtyp == 4;

[use_knots] = USER_INPUT_FINITE_CHOICE(...
    ['Speed options:\n[M] ' labelSpeed '\n[K] knots\nSelect option: '],...
    {'M', 'm', 'K', 'k'});
use_knots = strcmp(use_knots, 'K') || strcmp(use_knots, 'k');
if use_knots
    labelSpeedFinal = 'knots';
else
    labelSpeedFinal = labelSpeed;
end

% Single case input
if single_case
    [zobs] = USER_INPUT_DATA_VALUE(['Enter zobs: elevation of observed winds [' labelUnitDist ']: '], 1.0, 5000.0);

    [Uobs] = USER_INPUT_DATA_VALUE(['Enter uobs: observed wind speed [' labelSpeedFinal ']: '], 0.1, 200.0);

    [dtemp] = USER_INPUT_DATA_VALUE('Enter dtemp: air-sea temperature difference [deg C]: ', -100.0, 100.0);
    
    [duro] = USER_INPUT_DATA_VALUE('Enter duro: duration of observed wind [hr]: ', 0.1, 86400.0);
    
    [durf] = USER_INPUT_DATA_VALUE('Enter durf: duration of final wind [hr]: ', 0.1, 86400.0);
    
    [lat] = USER_INPUT_DATA_VALUE('Enter lat: latitude of wind observation [deg]: ', 0.0, 180.0);
    
    if is_water_open
        [F] = USER_INPUT_DATA_VALUE(['Enter F: length of wind fetch [' labelUnitDistLrg ']: '], 0.0, 9999.0);
    end
    
    if is_water_shallow
        [d] = USER_INPUT_DATA_VALUE(['Enter d: average depth of fetch [' labelUnitDist ']: '], 0.1, 10000.0);
    else
        d = 0;
    end
    
    if ~is_water_open
        [wdir] = USER_INPUT_DATA_VALUE('Enter wdir: wind direction [deg]: ', 0.0, 360.0);
    end
    
    numCases = 1;
else
    multiCaseData = {...
        ['zobs: elevation of observed winds [' labelUnitDist ']'], 1.0, 5000.0;...
        ['uobs: observed wind speed [' labelSpeedFinal ']'], 0.1, 200.0;...
        'dtemp: air-sea temperature difference [deg C]', -100.0, 100.0;...
        'duro: duration of observed wind [hr]', 0.1, 86400.0;...
        'durf: duration of final wind [hr]', 0.1, 86400.0;...
        'lat: latitude of wind observation [deg]', 0.0, 180.0};
    
    if is_water_open
        multiCaseData = [multiCaseData; {['F: length of wind fetch [' labelUnitDistLrg ']'], 0.0, 9999.0}];
    end
    
    if is_water_shallow
        multiCaseData = [multiCaseData; {['d: average depth of fetch [' labelUnitDist ']'], 0.1, 10000.0}];
    end
    
    if ~is_water_open
        multiCaseData = [multiCaseData;...
            {'wdir: wind direction [deg]', 0.0, 360.0}];
    end
    
    [varData, numCases] = USER_INPUT_MULTI_MODE(multiCaseData);
    
    zobsList = varData(1, :);
    uobsList = varData(2, :);
    dtempList = varData(3, :);
    duroList = varData(4, :);
    durfList = varData(5, :);
    latList = varData(6, :);
    
    optVarNum = 7;
    
    if is_water_open
        FList = varData(optVarNum, :);
        optVarNum = optVarNum + 1;
    end
    
    if is_water_shallow
        dList = varData(optVarNum, :);
        optVarNum = optVarNum + 1;
    end
    
    if ~is_water_open
        wdirList = varData(optVarNum, :);
    end
end

if ~is_water_open
    [dang] = USER_INPUT_DATA_VALUE('Enter dang: radial angle increment [deg]: ', 1.0, 180.0);

    [ang1] = USER_INPUT_DATA_VALUE('Enter ang1: direction of first radial fetch [deg]: ', 0.0, 360.0);

    [manualOrFile] = USER_INPUT_FINITE_CHOICE(...
        ['Would you like to enter fetch length data manually or load from a file?\n'...
            '[M] for manual entry or [F] for file loading: '],...
        {'M', 'm', 'F', 'f'});
    data_entry_manual = strcmp(manualOrFile, 'M') || strcmp(manualOrFile, 'm');

    if data_entry_manual
        [Nfet] = USER_INPUT_DATA_VALUE('Enter Nfet: number of radial fetches: ', 2, 360);
        Nfet = floor(Nfet);

        angs = [];
        for angsLoopIndex = 1:Nfet
            angTemp = USER_INPUT_DATA_VALUE(['Enter angs: fetch length [' labelUnitDistLrg '] #' num2str(angsLoopIndex) ': '], 0,9999);

            angs = [angs angTemp];
        end

        clear angsLoopIndex;
        clear angTemp;
    else
        accepted = false;
        while ~accepted
            [filename] = USER_INPUT_FILE_NAME();

            fId = fopen(filename);

            fileData = textscan(fId, '%f');

            fclose(fId);

            if length(fileData{1}) >= 2 && length(fileData{1}) <= 360
                accepted = true;
            else
                fprintf('File must have 2 to 360 fetch lengths.\n');
            end
        end

        angs = fileData{1};
        Nfet = length(angs);
    end
end


% Constant for convertions
ft2m=0.3048;
mph2mps=0.44704;
hr2s=3600;
min2s=60;
deg2rad=pi/180;
mi2m=1609.344;
km2m=0.001;
F2C=5/9;
knots2mps=0.5144;

if ~metric
    conversionDist = ft2m;
    conversionDistLrg = mi2m;
else
    conversionDist = 1.0;
    conversionDistLrg = km2m;
end

if use_knots
    conversionSpeed = knots2mps;
elseif ~metric
    conversionSpeed = mph2mps;
else
    conversionSpeed = 1.0;
end

% File Output
fileOutputArgs = {};
[fileOutputData] = USER_INPUT_FILE_OUTPUT(fileOutputArgs);

if fileOutputData{1}
    fId = fopen('output/wind_adj.txt', 'wt');
    
    fprintf(fId, '%s\n', windObsList{windobs});
    fprintf(fId, '%s\n\n', wgTypeList{wgtyp});
end

for loopIndex = 1:numCases
    if ~single_case
        zobs = zobsList(loopIndex);
        Uobs = uobsList(loopIndex);
        dtemp = dtempList(loopIndex);
        duro = duroList(loopIndex);
        durf = durfList(loopIndex);
        lat = latList(loopIndex);
        
        if is_water_open
            F = FList(loopIndex);
        end
        
        if is_water_shallow
            d = dList(loopIndex);
        else
            d = 0;
        end
        
        if ~is_water_open
            wdir = wdirList(loopIndex);
        end
    end
    
    errorMsg = '';
    
    if is_water_open
        phi=0;
    else
        [F,phi,theta]=WGFET(ang1,dang,wdir,angs);
    end

%     assert(lat~=0, 'Error: Latitude must be a non-zero value.')
    if lat == 0
        errorMsg = 'Error: Latitude must be a non-zero value.';
        disp(errorMsg);
    else
    %   Check WDIR vs Fetch data. WDIR must meet this criterion:
    %   ang1 -45 degrees <= WDIR <= anglast + 45 degrees
        if ~is_water_open
%             assert(ang1 - 45 <= wdir,...
%                 'Error: wdir must be at least 45 degrees less than the first fetch angle.');
%             assert(wdir <= (ang1 + (Nfet - 1)*dang) + 45,...
%                 'Error: wdir must be at most 45 degrees more than the final fetch angle.');
            if not(ang1 - 45 <= wdir)
                errorMsg = 'Error: wdir must be at least 45 degrees less than the first fetch angle.';
                disp(errorMsg);
            else
                if not(wdir <= (ang1 + (Nfet - 1)*dang) + 45)
                    errorMsg = 'Error: wdir must be at most 45 degrees more than the final fetch angle.';
                    disp(errorMsg);
                end
            end
        end
    end
    
    if length(errorMsg) == 0
    %    [ue]=WADJ(Uobs*mph2mps,zobs*ft2m,dtemp,F*mi2m,duro*hr2s,durf*hr2s,lat*deg2rad,windobs);
        [ue]=WADJ(Uobs*conversionSpeed,...
            zobs*conversionDist,...
            dtemp,...
            F*conversionDistLrg,...
            duro*hr2s,...
            durf*hr2s,...
            lat*deg2rad,...
            windobs);

        [ua,Hmo,Tp,wgmsg]=WGRO(d*conversionDist,F*conversionDistLrg,phi,durf*hr2s,ue,wgtyp);

        if ~is_water_open
            fprintf('%s \t\t\t\t %-6.2f %s\n','Fetch length',F, labelUnitDistLrg);
            fprintf('%s \t\t\t %-6.2f deg\n', 'Wind Direction', wdir);
        end

    %    fprintf('%s \t\t %-6.2f %s \n','Equiv. wind speed',ue/mph2mps, labelSpeedFinal);
        fprintf('%s \t\t %-6.2f %s \n','Equiv. wind speed',ue/conversionSpeed, labelSpeedFinal);
    %    fprintf('%s \t\t %-6.2f %s \n','Adjus. wind speed',ua/mph2mps, labelSpeedFinal);
        fprintf('%s \t\t %-6.2f %s \n','Adjus. wind speed',ua/conversionSpeed, labelSpeedFinal);

        if ~is_water_open
            fprintf('%s \t %-6.2f deg\n','Mean wave direction',theta);
        end

        fprintf('%s \t\t\t %-6.2f %s \n','Wave height ',Hmo/conversionDist,labelUnitDist);
        fprintf('%s \t\t\t %-6.2f s \n','Wave period ',Tp);

        fprintf('%s %s \n','Wave growth: ',wgmsg);
    end
    
    if fileOutputData{1}
        if ~single_case
            fprintf(fId, 'Case #%d\n\n', loopIndex);
        end
        
        fprintf(fId, 'Input\n');
        fprintf(fId, 'zobs\t\t%6.2f %s\n', zobs, labelUnitDist);
        fprintf(fId, 'uobs\t\t%6.2f %s\n', Uobs, labelSpeedFinal);
        fprintf(fId, 'dtemp\t\t%6.2f deg\n', dtemp);
        fprintf(fId, 'duro\t\t%6.2f hr\n', duro);
        fprintf(fId, 'durf\t\t%6.2f hr\n', durf);
        fprintf(fId, 'lat\t\t%6.2f deg\n', lat);
        
        if is_water_open
            fprintf(fId, 'F\t\t%6.2f %s\n', F, labelUnitDistLrg);
        end
        
        if is_water_shallow
            fprintf(fId, 'd\t\t%6.2f %s\n', d, labelUnitDist);
        end
        
        if ~is_water_open
            fprintf(fId, 'wdir\t\t%6.2f deg\n', wdir);
        end

        if length(errorMsg) > 0
            fprintf(fId, '\n%s\n', errorMsg);
        else
            fprintf(fId, '\nOutput\n');

            if ~is_water_open
                fprintf(fId, '%s \t\t\t %-6.2f %s\n','Wind fetch',F, labelUnitDistLrg);
                fprintf(fId, '%s \t\t\t %-6.2f deg\n', 'Wind Direction', wdir);
            end

            fprintf(fId, '%s \t\t %-6.2f %s \n','Equiv. wind speed',ue/conversionSpeed, labelSpeedFinal);
            fprintf(fId, '%s \t\t %-6.2f %s \n','Adjus. wind speed',ua/conversionSpeed, labelSpeedFinal);

            if ~is_water_open
                fprintf(fId, '%s \t\t %-6.2f deg\n','Mean wave direction',theta);
            end

            fprintf(fId, '%s \t\t\t %-6.2f %s \n','Wave height ',Hmo/conversionDist,labelUnitDist);
            fprintf(fId, '%s \t\t\t %-6.2f s \n','Wave period ',Tp);

            fprintf(fId, '%s %s \n','Wave growth: ',wgmsg);
        end
        
        if loopIndex < numCases
            fprintf(fId, '\n--------------------------------------\n\n');
        end
    end
end

if fileOutputData{1}
    fclose(fId);
end

File: python/drivers/wind_adj.py
import sys
import math
import numpy as np
sys.path.append('../functions')

from base_driver import BaseDriver
from helper_objects import BaseField
import USER_INPUT
from WADJ import WADJ
from WGFET import WGFET
from WGRO import WGRO

## ACES Update to python
#-------------------------------------------------------------
# Driver for Windspeed Adjustment and Wave Growth (page 1-1 of ACES User's
# Guide). Provide estimates for wave growth over open-water and restricted
# fetches in deep and shallow water.

# Updated by: Mary Anderson, USACE-CHL-Coastal Processes Branch
# Date Created: April 28, 2011
# Date Modified:

# Requires the following functions:
# ERRWAVBRK1
# WADJ
# WAGEOS
# WAPBL
# WAPSI
# WASBL
# WASHR
# WGDL
# WGFD
# WGFET
# WGFL
# WGRO

# MAIN VARIABLE LIST:
#   INPUT
#   zobs: elevation of observed winds [m]
#   uobs: observed wind speed [m/s]
#   dtemp: air-sea temperature difference [deg C]
#   duro: duration of observed wind [hr]
#   durf: duration of final wind [hr]
#   lat: latitude of wind observation [deg]
#   windobs: wind observation type
#   fetchopt: wind fetch options
#   wgtyp: open water wave growth equation options

#   OPEN-WATER VARIABLES
#   F: length of wind fetch [m]
#   d: average depth of fetch (only for shallow water equations) [m]

#   RESTRICTED VARIABLES
#   wdir: wind direction [deg]
#   dang: radial angle increment [deg]
#   ang1: direction of first radial fetch [deg]
#   angs: fetch length [m]

#   OUTPUT
#   ue: equivalent neutral wind speed [m/s]
#   ua: adjusted wind speed [m/s]
#   Hmo: wave height [m]
#   Tp: peak wave period [s]
#   wg: type of wave-growth
#   theta: wave direction with respect to N [deg]

#   OTHERS
#-------------------------------------------------------------

class WindAdj(BaseDriver):
    def __init__(self, windobs = None, wgtyp = None, useKnots = None,\
        zobs = None, Uobs = None, dtemp = None, duro = None, durf = None,\
        lat = None, F = None, d = None, wdir = None, dang = None,\
        ang1 = None, manualOrFile = None, Nfet = None, angs = None):
        if windobs != None:
            self.isSingleCase = True
            self.defaultValue_windobs = windobs
        if wgtyp != None:
            self.isSingleCase = True
            self.defaultValue_wgtyp = wgtyp
        if useKnots != None:
            self.isSingleCase = True
            self.defaultValue_useKnots = useKnots
        if zobs != None:
            self.isSingleCase = True
            self.defaultValue_zobs = zobs
        if Uobs != None:
            self.isSingleCase = True
            self.defaultValueUobs = Uobs
        if dtemp != None:
            self.isSingleCase = True
            self.defaultValue_dtemp = dtemp
        if duro != None:
            self.isSingleCase = True
            self.defaultValue_duro = duro
        if durf != None:
            self.isSingleCase = True
            self.defaultValue_durf = durf
        if lat != None:
            self.isSingleCase = True
            self.defaultValue_lat = lat
        if F != None:
            self.isSingleCase = True
            self.defaultValueF = F
        if d != None:
            self.isSingleCase = True
            self.defaultValue_d = d
        if wdir != None:
            self.isSingleCase = True
            self.defaultValue_wdir = wdir
        if dang != None:
            self.isSingleCase = True
            self.defaultValue_dang = dang
        if ang1 != None:
            self.isSingleCase = True
            self.defaultValue_ang1 = ang1
        if manualOrFile != None:
            self.isSingleCase = True
            self.defaultValue_manualOrFile = manualOrFile
        if Nfet != None:
            self.isSingleCase = True
            self.defaultValue_useKnots = useKnots
        if angs != None:
            self.isSingleCase = True
            self.defaultValue_useKnots = useKnots

        super(WindAdj, self).__init__()
    # end __init__

    def userInput(self):
        self.windObsList = ["Overwater (shipboard)",\
            "Overwater (not shipboard)",\
            "Shore (windward - offshore to onshore)",\
            "Shore (leeward - onshore to offshore)",\
            "Over land",\
            "Geostrophic wind"]

        if not hasattr(self, "defaultValue_windobs"):
            windObsMsg = "Wind observation types:\n"
            for i in range(len(self.windObsList)):
                windObsMsg += "[%d] %s\n" % ((i + 1), self.windObsList[i])
            windObsMsg += "Select option: "

            self.windobs = USER_INPUT.FINITE_CHOICE(\
                windObsMsg, ["1", "2", "3", "4", "5", "6"])
            self.windobs = int(self.windobs)
        else:
            self.windobs = self.defaultValue_windobs

        self.wgtypList = ["Open Water - Deep",\
            "Open Water - Shallow",\
            "Restricted - Deep",\
            "Restricted - Shallow"]

        if not hasattr(self, "defaultValue_wgtyp"):
            wgtypMsg = "Wind fetch and wave growth options:\n"
            for i in range(len(self.wgtypList)):
                wgtypMsg += "[%d] %s\n" % ((i + 1), self.wgtypList[i])
            wgtypMsg += "Select option: "

            self.wgtyp = USER_INPUT.FINITE_CHOICE(\
                wgtypMsg, ["1", "2", "3", "4"])
            self.wgtyp = int(self.wgtyp)
        else:
            self.wgtyp = self.defaultValue_wgtyp

        self.isWaterOpen = self.wgtyp == 1 or self.wgtyp == 2
        self.isWaterShallow = self.wgtyp == 2 or self.wgtyp == 4

        super(WindAdj, self).userInput()

        if not self.isWaterOpen:
            if not hasattr(self, "defaultValue_dang"):
                self.dang = USER_INPUT.DATA_VALUE(\
                    "dang: radial angle increment [deg]", 1.0, 180.0)
            else:
                self.dang = self.defaultValue_dang

            if not hasattr(self, "defaultValue_ang1"):
                self.ang1 = USER_INPUT.DATA_VALUE(\
                    "ang1: direction of first radial fetch [deg]",\
                    0.0, 360.0)
            else:
                self.ang1 = self.defaultValue_ang1

            if not hasattr(self, "defaultValue_manualOrFile"):
                manualOrFile = USER_INPUT.FINITE_CHOICE(\
                    "Would you like to enter fetch length data " +\
                    "manually or load from a file?\n" +\
                    "[M] for manual entry or [F] for file loading: ",
                    ["M", "m", "F", "f"])

                if manualOrFile == "M" or manualOrFile == "m":
                    self.Nfet = USER_INPUT.DATA_VALUE(\
                        "Nfet: number of radial fetches", 2, 360)
                    self.Nfet = int(self.Nfet)

                    self.angs = []
                    for i in range(self.Nfet):
                        self.angs.append(USER_INPUT.DATA_VALUE(\
                            "angs: fetch length [%s] #%d" %\
                            (self.labelUnitDistLrg, i), 0.0, 9999.0))
                else:
                    accepted = False
                    while not accepted:
                        filename = USER_INPUT.FILE_NAME()

                        fileRef = open(filename)

                        fileData = fileRef.read()

                        self.angs = []
                        for dataLine in fileData.split("\n"):
                            if len(dataLine) > 0:
                                self.angs.append(float(dataLine))
                        self.Nfet = len(self.angs)

                        fileRef.close()

                        if self.Nfet >= 2 and self.Nfet <= 360:
                            accepted = True
                        else:
                            print("File must have between 2 and 360 fetch lengths.")
                    # end while
    # end userInput

    def defineInputDataList(self):
        if self.isMetric:
            self.labelSpeed = "m/s"
            self.labelUnitDistLrg = "km"
        else:
            self.labelSpeed = "mph"
            self.labelUnitDistLrg = "mi"

        if not hasattr(self, "defaultValue_useKnots"):
            self.useKnots = USER_INPUT.FINITE_CHOICE(\
                "Speed options:\n[M] %s\n[K] knots\nSelect option: " %\
                self.labelSpeed,\
                ["M", "m", "K", "k"])
        else:
            self.useKnots = self.defaultValue_useKnots
        if self.useKnots == "K" or self.useKnots == "k":
            self.labelSpeedFinal = "knots"
            self.useKnots = True
        else:
            self.labelSpeedFinal = self.labelSpeed
            self.useKnots = False

        self.inputList = []

        if not hasattr(self, "defaultValue_zobs"):
            self.inputList.append(BaseField(\
                "zobs: elevation of observed winds [%s]" % (self.labelUnitDist),\
                1.0, 5000.0))
        if not hasattr(self, "defaultValueUobs"):
            self.inputList.append(BaseField(\
                "uobs: observed wind speed [%s]" % self.labelSpeedFinal,\
                0.1, 200.0))
        if not hasattr(self, "defaultValue_dtemp"):
            self.inputList.append(BaseField(\
                "dtemp: air-sea temperature difference [deg C]",\
                -100.0, 100.0))
        if not hasattr(self, "defaultValue_duro"):
            self.inputList.append(BaseField(\
                "duro: duration of observed wind [hr]", 0.1, 86400.0))
        if not hasattr(self, "defaultValue_durf"):
            self.inputList.append(BaseField(\
                "durf: duration of final wind [hr]", 0.1, 86400.0))
        if not hasattr(self, "defaultValue_lat"):
            self.inputList.append(BaseField(\
                "lat: latitude of wind observation [deg]", 0.0, 180.0))

        if self.isWaterOpen and not hasattr(self, "defaultValueF"):
            self.inputList.append(BaseField(\
                "F: length of wind fetch [%s]" % self.labelUnitDistLrg,
                0.0, 9999.0))

        if self.isWaterShallow and not hasattr(self, "defaultValue_d"):
            self.inputList.append(BaseField(\
                "d: average depth of fetch [%s]" % self.labelUnitDist,
                0.1, 10000.0))

        if not self.isWaterOpen and not hasattr(self, "defaultValue_wdir"):
            self.inputList.append(BaseField(\
                "wdir: wind direction [deg]", 0.0, 360.0))
    # end defineInputDataList

    def fileOutputRequestInit(self):
        self.fileOutputRequestMain(defaultFilename = "wind_adj")

    def getCalcValues(self, caseInputList):
        currIndex = 0

        if hasattr(self, "defaultValue_zobs"):
            zobs = self.defaultValue_zobs
        else:
            zobs = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValueUobs"):
            Uobs = self.defaultValueUobs
        else:
            Uobs = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_dtemp"):
            dtemp = self.defaultValue_dtemp
        else:
            dtemp = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_duro"):
            duro = self.defaultValue_duro
        else:
            duro = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_durf"):
            durf = self.defaultValue_durf
        else:
            durf = caseInputList[currIndex]
            currIndex = currIndex + 1

        if hasattr(self, "defaultValue_lat"):
            lat = self.defaultValue_lat
        else:
            lat = caseInputList[currIndex]
            currIndex = currIndex + 1

        if self.isWaterOpen:
            if hasattr(self, "defaultValueF"):
                F = self.defaultValueF
            else:
                F = caseInputList[currIndex]
                currIndex = currIndex + 1
        else:
            F = None

        if self.isWaterShallow:
            if hasattr(self, "defaultValue_d"):
                d = self.defaultValue_d
            else:
                d = caseInputList[currIndex]
                currIndex = currIndex + 1
        else:
            d = 0.0

        if not self.isWaterOpen:
            if hasattr(self, "defaultValue_wdir"):
                wdir = self.defaultValue_wdir
            else:
                wdir = caseInputList[currIndex]
                currIndex = currIndex + 1
        else:
            wdir = None

        if not self.isWaterOpen:
            dang = self.dang
            ang1 = self.ang1
            Nfet = self.Nfet
            angs = self.angs
        else:
            dang = None
            ang1 = None
            Nfet = None
            angs = None

        return zobs, Uobs, dtemp, duro, durf, lat, F, d, wdir,\
            dang, ang1, Nfet, angs
    # end getCalcValues

    def performCalculations(self, caseInputList, caseIndex = 0):
        zobs, Uobs, dtemp, duro, durf, lat, F, d, wdir,\
            dang, ang1, Nfet, angs = self.getCalcValues(caseInputList)

        # Constant for convertions
        ft2m = 0.3048
        mph2mps = 0.44704
        hr2s = 3600.0
        min2s = 60.0
        deg2rad = math.pi/180.0
        mi2m = 1609.344
        km2m = 0.001
        F2C = 5.0/9.0
        knots2mps = 0.5144

        if not self.isMetric:
            conversionDist = ft2m
            conversionDistLrg = mi2m
        else:
            conversionDist = 1.0
            conversionDistLrg = km2m

        if self.useKnots:
            conversionSpeed = knots2mps
        elif not self.isMetric:
            conversionSpeed = mph2mps
        else:
            conversionSpeed = 1.0

        if self.isWaterOpen:
            phi = 0.0
        else:
            F, phi, theta = WGFET(ang1, dang, wdir, angs)

        if np.isclose(lat, 0.0):
            print("Error: Latitude must be a non-zero value.")
            return

        # Check WDIR vs Fetch data. WDIR must meet this criterion:
        # ang1 -45 degrees <= WDIR <= anglast + 45 degrees
        if not self.isWaterOpen:
            if not (ang1 - 45 <= wdir):
                print("Error: wdir must be at least 45 degrees less than the first fetch angle.")
                return
            if not (wdir <= (ang1 + (Nfet - 1)*dang)):
                print("Error: wdir must be at most 45 degrees more than the final fetch angle.")
                return

        ue = WADJ(Uobs*conversionSpeed,\
            zobs*conversionDist,\
            dtemp,\
            F*conversionDistLrg,\
            duro*hr2s,\
            durf*hr2s,\
            lat*deg2rad,\
            self.windobs)

        ua, Hmo, Tp, wgmsg = WGRO(d*conversionDist,\
            F*conversionDistLrg,\
            phi,
            durf*hr2s,\
            ue,\
            self.wgtyp)

        dataDict = {"zobs": zobs, "Uobs": Uobs, "dtemp": dtemp,\
            "duro": duro, "durf": durf, "lat": lat, "F": F,\
            "d": d, "wdir": wdir, "dang": dang, "ang1": ang1}
        self.fileOutputWriteMain(dataDict, caseIndex)
    # end performCalculations

    def fileOutputWriteData(self, dataDict):
        self.fileRef.write("Input\n")
        self.fileRef.write("zobs\t%6.2f %s\n" % (dataDict["zobs"], self.labelUnitDist))
        self.fileRef.write("uobs\t%6.2f %s\n" % (dataDict["Uobs"], self.labelSpeedFinal))
        self.fileRef.write("dtemp\t%6.2f deg\n" % dataDict["dtemp"])
        self.fileRef.write("duro\t%6.2f hr\n" % dataDict["duro"])
        self.fileRef.write("durf\t%6.2f hr\n" % dataDict["durf"])
        self.fileRef.write("lat\t%6.2f deg\n" % dataDict["lat"])

        if self.isWaterOpen:
            self.fileRef.write("F\t%6.2f %s\n" % (dataDict["F"], self.labelUnitDistLrg))

        if self.isWaterShallow:
            self.fileRef.write("d\t%6.2f %s\n" % (dataDict["d"], self.labelUnitDist))

        if not self.isWaterOpen:
            self.fileRef.write("wdir\t%6.2f deg\n" % dataDict["wdir"])

        self.fileRef.write("\nOutput\n")
    # end fileOutputWriteData


driver = WindAdj(zobs = 30.0, Uobs = 45.0, dtemp = -3.0, duro = 5.0,\
    durf = 5.0, lat = 47.0, wdir = 125.0, dang = 12.0, ang1 = 0.0)